src/gnutls.c

```diff
@@ -1049,33 +1050,41 @@ client = INVALID_POOL_HANDLE;
 			tlog (TLOG_COPYCAT, LOG_DEBUG, "Copycat polling returned an error");
 			break;	// Polling sees an error
 		}
-		if (inout [0].revents & POLLIN) {
+		tlog (TLOG_COPYCAT, LOG_DEBUG, "inout [0].revents 0x%04x", inout [0].revents);
+		tlog (TLOG_COPYCAT, LOG_DEBUG, "inout [1].revents 0x%04x", inout [1].revents);
+		if (inout [0].revents & (POLLIN | POLLHUP)) {
 			// Read local and encrypt to remote
-			sz = recv (local, buf, sizeof (buf), RECV_FLAGS);
+			sz = (inout [0].revents & POLLIN) || !local_shutdown_done ? recv (local, buf, sizeof (buf), RECV_FLAGS) : 0;
 			tlog (TLOG_COPYCAT, LOG_DEBUG, "Copycat received %d local bytes (or error<0) from %d", (int) sz, local);
 			if (sz == -1) {
 				tlog (TLOG_COPYCAT, LOG_ERR, "Error while receiving: %s", strerror (errno));
 				break;	// stream error
 			} else if (sz == 0) {
 				inout [0].events &= ~POLLIN;
-				shutdown (local, SHUT_RD);
+				if (!local_shutdown_done) {
+					tlog (TLOG_COPYCAT, LOG_DEBUG, "Sending gnutls_bye");
+					shutdown (local, SHUT_RD);
 #ifdef WINDOWS_PORT
```

Comment:
is de enige reden voor deze `#ifdef WINDOWS_PORT` het type?  dan zouden we kunnen overwegen naar (void *) te casten voor beide en het onderscheid te verwijderen?  het is toch maar een type dat substituut is voor een vrijheid tot samenkomen die de jaren 60 zou sieren...